src/System.IO.FileSystem/src/System/IO/Win32FileSystemEnumerable.cs

```diff
@@ -192,39 +198,38 @@ private void CommonInit()
                 }
                 else
                 {
-                    SearchResult searchResult = CreateSearchResult(_searchData, data);
-                    if (_resultHandler.IsResultIncluded(searchResult))
+                    TSource result;
+                    if (IsResultIncluded(ref data, out result))
                     {
-                        current = _resultHandler.CreateObject(searchResult);
+                        current = result;
                     }
                 }
             }
-            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData 
+            // for AllDirectories, we first recurse into dirs, so cleanup and add searchData
             // to the stack
             else
             {
                 _hnd.Dispose();
+                _searchStack = new List<SearchData>();
```

Comment:
Do you know why we're using a ```List<T>``` instead of a ```Stack<T>```?  With most of the operations happening at the 0th element of the list, we're going to incur a bunch of unnecessary moves in the list to continually shift elements around.

Seems like the only non-stack-ish operation being performed is that when iterating through a directory, it looks like we're maintaining a FIFO order of the elements added, just for that set.  If that's not important, seems we should just switch to using a ```Stack<T>```.  If that is important, we could stick with a ```List<T>``` but at least operate at the other end of the list, to avoid all of those shift operations; in the case of the FIFO ordering, we could still just Add everything but then Reverse just the portion that was added to restore that order.

Note that by changing SearchData from a class to a struct, we've changed the amount of data stored in the list, such that each element is now the size of two references rather than just one, so moves will become more expensive.